### DISCUSSION TITLE
How do you post `FormData` with a duplex (or transform) stream?

### DISCUSSION THREAD

**Discussion by kiyutink** [sindresorhus/got#1674] [Q&A] [answered]
I was trying to fetch a file from a remote URL and upload it to a different url without saving it to the disk first. That wasn't working and i was getting a `RequestError`.  I did some poking around and noticed that if i use a read stream passed through a simple transform stream, i was getting the same error. 
Could someone give me a pointer to what's going wrong? Is it me doing something wrong or the library doesn't support this? If latter, is it something that can be implemented in theory?

```js

// working:

const readStream = fs.createReadStream("file.txt");
const form = new FormData();
form.append("file", readStream);
got.post("https://destination.url", {
  body: form
});


// not working:
const remoteFileStream = got.stream("https://remote-file.url")
const form = new FormData();
form.append("file", remoteFileStream);
got.post("https://destination.url", {
  body: form
});


// not working:

const readStream = fs.createReadStream("file.txt");
const transformStream = new Transform({
  transform(chunk, encoding, callback) {
    callback(null, chunk);
  },
});
const form = new FormData();
form.append("file", readStream.pipe(transformStream));
got.post("https://destination.url", {
  body: form
});



```

**Comment by zzlPetrichor** [comment]
I have the same question.
Did you solve it? please.

**Reply by kiyutink** [reply]
i wasn't able to pipe it directly as I wanted, IIRC this limitation is imposed by the "form-data" library
For me the streaming wasn't critical as i was working with small files, so i just loaded the entire file in memory as buffer and sent it over from memory

**Reply by zzlPetrichor** [reply]
thank you.
I just solve it.
first, get the response from stream with on event api.
then, replace readStream with response in form-data.
And I find it only works when the got version is over 11.0.0+. (important)

**Comment by hctan** [comment]
hi @zzlPetrichor , would you be able to share the code snippet how you got it to work? I tried the same way (passing in response) but unable to get it to work. Here's my code:

    ```
    const got = require('got');
    const FormData = require('form-data');

    const req = got.stream('https://sourceUrl');

    req.on('response', async (res) => {
        res.pause();

        const formData = new FormData();
        formData.append('file', res, { filename: 'file1.txt', knownLength: NaN });

        try
        {
            await got.post('https://destinationUrl', {
                body: formData,
            });
        } catch (ex) {
        }
    });
```

I'll get a message saying its response code 408 when it tries to upload the file (got.post()) but I can confirm that the destination server didn't receive the call. It works if I I buffer the stream first and pass the buffer to formData's file.

**Comment by sindresorhus** [comment, marked as answer]
Posting FormData with streams can be tricky because FormData needs to know the content length upfront. Here's the recommended approach:

```javascript
import got from 'got';
import FormData from 'form-data';

// Fetch the file as a stream
const sourceStream = got.stream('https://example.com/file.pdf');

// Wait for response event to get headers (including content-length)
sourceStream.on('response', (response) => {
	const form = new FormData();
	
	form.append('file', sourceStream, {
		filename: 'file.pdf',
		contentType: response.headers['content-type'],
		knownLength: parseInt(response.headers['content-length'], 10)
	});
	
	// Upload the form
	got.post('https://upload.example.com', {
		body: form
	});
});
```

The key is providing `knownLength` so FormData can set the proper `Content-Length` header. Without it, the upload may fail or timeout.

If the source doesn't provide a content-length header, you'll need to buffer the stream first, which defeats the puprose of streaming.
